### PR TITLE
feat: Add `HttpRequest::timeout_ms` and `Manifest::http_timeout_ms` 

### LINKED ISSUE #151: Configurable HTTP timeouts
For `extism_http_request` in particular, we should add both host-configurable HTTP timeouts for call-outs made from a plug-in, as well as a way to set it from the PDK, but not to exceed one if set by the host. The PDK should only be able to set a timeout that is shorter than one set by the host.

### PR DESCRIPTION
**Pull request by zshipko**
Closes #151 

- Adds ability to configure timeout for `extism_http_request`
- Sets default timeout to 30s, but this can be overridden still
- Updates SDKs to support the new fields

### CODE REVIEW

File: runtime/src/pdk.rs [hunk 1]
```diff
@@ -338,6 +338,19 @@ pub(crate) fn http_request(
             r = r.set(k, v);
         }
 
+        // Configure timeout
+        let global_timeout = data.plugin().manifest.as_ref().http_timeout_ms;
+        let req_timeout = req.timeout_ms;
+
+        r = r.timeout(std::time::Duration::from_micros(
```
Comment (nilslice) [inline runtime/src/pdk.rs:345]: is `from_micros` here intentional?

File: runtime/src/manifest.rs [hunk 2]
```diff
@@ -142,6 +143,10 @@ fn to_module(engine: &Engine, wasm: &extism_manifest::Wasm) -> Result<(String, M
                     req = req.set(k, v);
                 }
 
+                req = req.timeout(std::time::Duration::from_micros(
```
Comment (nilslice) [inline runtime/src/manifest.rs:146]: also here, `from_micros` -- should this be `from_millis` ?

File: runtime/src/pdk.rs (hunk 1 above)
Comment (zshipko) [reply runtime/src/pdk.rs:345]: No, thanks for catching that!

Comment (nilslice) [reply runtime/src/pdk.rs]: pushed up a fix!

**Comment by nilslice** [comment]
Testing this out now, all is working great - just curious about a couple things: 

1. for a host that wants to override the default to effectively disable a timeout, letting a plugin wait for a response indefinitely, I believe the only way to do that would be to use [`Duration::MAX`](https://doc.rust-lang.org/std/time/struct.Duration.html#associatedconstant.MAX)? I think that's reasonable.. and we can document that. Let me know if that's untrue though!
2. it seems as though there is no way for a plugin to recover from a host-imposed timeout. I can see how this would be useful as a control mechanism from the host, but think that we should instead lift that responsibility to a more general plugin timeout, and not cause an HTTP timeout to bail the whole plugin. 

re: 2, I'm looking at the implementation though and it seems like the `http_request` PDK function should just return the error to the caller within the plugin, but from my host, I can't get anything back except an err:

host:
```rust
use extism::*;

fn main() -> Result<(), Error> {
    let context = Context::new();
    let manifest = Manifest::new([std::path::PathBuf::from("code.wasm")])
        .with_http_timeout(std::time::Duration::from_millis(1000));

    let mut plugin = PluginBuilder::new(manifest).build(&context)?;

    match plugin.call("test", &[]) {
        Err(e) => println!("err: {:?}", e),
        Ok(s) => println!("output: {}", String::from_utf8(s.to_vec()).unwrap()),
    }

    return Ok(());
}
```

plugin: 

```rust
use extism_pdk::*;

#[plugin_fn]
pub fn test(_input: ()) -> FnResult<String> {
    let req = HttpRequest::new("https://httpstat.us/200?sleep=2000");
    match http::request::<()>(&req, None) {
        Err(_) => Ok(format!("error!! but sending Result::Ok")),
        Ok(res) => Ok(format!("status: {}", res.status_code())),
    }
}
```

**Note**: both match arms in the plugin return `Ok`, so the runtime is propagating the error up before the plugin can do anything about it.

**Comment by zshipko** [comment]
I think setting timeout to 0 will also wait indefinitely, I thought I saw that in the docs but can’t find it right now.

I can test that out and look into the error propagation issue tomorrow - I have a pretty good idea of how to fix that and I will update this PR.

**Comment by nilslice** [comment]
Ok sounds good! Thanks. I'll also take a look to see about the 0 timeout setting in the http client.

**Comment by nilslice** [comment]
update: setting the timeout to `0` bails immediately with the timeout error

**Comment by zshipko** [comment]
Okay, thanks for checking! Looks like `Duration::MAX` is the best option then.

**Comment by bhelx** [comment]
We're closing this now that we have a top level function timeout. If anyone does feel we need to support this, we can re-open.

**Pull request closed without merging**